src/charm.py

```diff
@@ -219,25 +279,130 @@ def _generate_kiali_layer() -> Layer:
+
+        Returns None if we are not related to a tempo.
+
+        Raises:
+          ConfigurationBlockingError: If a required relation for configuring Tempo is missing.
+          ConfigurationWaitingError: If a Tempo is related to this application, but its data is incomplete.
+        """
+        try:
+            tempo_metadata = self._get_tempo_metadata()
+        except ConfigurationWaitingError as e:
+            raise ConfigurationWaitingError(f"Tempo metadata is incomplete.  Got error: {e}")
+
+        if not tempo_metadata:
+            # We have no tempo related to us
+            return None
+
+        try:
+            tempo_datasource_uid = self._get_tempo_datasource_uid()
+        except ConfigurationBlockingError as e:
+            raise ConfigurationBlockingError(f"Tempo datasource data is missing.  Got error: {e}")
+        except ConfigurationWaitingError as e:
+            raise ConfigurationWaitingError(f"Tempo datasource data is incomplete, possibly because are waiting on "
+                                              f"the related application.  Got error: {e}")
+
+        return TempoConfigurationData(
+            internal_url=tempo_metadata.internal_url,
+            external_url=tempo_metadata.ingress_url,
+            grpc_port=tempo_metadata.grpc_port,
+            datasource_uid=tempo_datasource_uid,
+        )
+
+    def _get_tempo_metadata(self):
+        """Get the Tempo source urls (internal and external).
+
+        Raises:
+            ConfigurationWaitingError: If a Tempo is related to this application, but its data is incomplete.
+        """
+        # TODO: Use the real data type here
+        if len(self._tempo_metadata) == 0:
+            return None
+
+        tempo_metadata = self._tempo_metadata.get_data()
+        if not tempo_metadata:
+            raise ConfigurationWaitingError("Waiting on related tempo application's metadata")
+
+        return tempo_metadata
+
+    def _get_tempo_datasource_uid(self):
+        """Get the Tempo datasource uid.
+
+        Returns the first related datasource that is a tempo datasource and has the same grafana uid as the known
+        grafana.
+
+        Will raise:
+          ConfigurationBlockingError: If no applications are related, or applications are related but have sent only
+                                       non-tempo datasources
+          ConfigurationWaitingError: If a datasource is related, but has not yet provided data
+        """
+        try:
+            grafana_source_uid = self._get_grafana_uid()
+        except ConfigurationBlockingError as e:
+            raise ConfigurationBlockingError(f"Tempo datasource is missing because Grafana source is missing.  Got"
+                                              f" error: {e}")
+        except ConfigurationWaitingError as e:
+            raise ConfigurationWaitingError(f"Tempo datasource is missing because Grafana source is incomplete.  Got"
```

Comment:
nit: no need to include the parent error in the message: it's already included in the traceback. IMO here we should:
```suggestion
            raise ConfigurationWaitingError(f"Tempo datasource is missing")"
```
the rest of the traceback will give you the 'because'

Reply:
I agree with you if we're letting an exception surface this makes sense (probably we'd add a `raise X from e` here too), but in this case the Exception's message is being used as the WaitingStatus's message.  The traceback will be lost

This does feel like a slight abuse of exceptions though

Reply:
yep it does

Reply:
you could write the exception handler in the collect-status to collate the messages of the whole exception stack though

Reply:
I don't see that making things clearer for a user or easier to develop for a developer